mount/mount_unix.go

```diff
@@ -18,24 +18,41 @@
 
 package mount
 
-import "errors"
+import (
+	"sort"
 
-var (
-	// ErrNotImplementOnUnix is returned for methods that are not implemented
-	ErrNotImplementOnUnix = errors.New("not implemented under unix")
+	"github.com/moby/sys/mountinfo"
 )
 
-// Mount is not implemented on this platform
-func (m *Mount) Mount(target string) error {
-	return ErrNotImplementOnUnix
-}
-
-// Unmount is not implemented on this platform
-func Unmount(mount string, flags int) error {
-	return ErrNotImplementOnUnix
-}
-
-// UnmountAll is not implemented on this platform
-func UnmountAll(mount string, flags int) error {
-	return ErrNotImplementOnUnix
+// UnmountRecursive unmounts the target and all mounts underneath, starting
+// with the deepest mount first.
+func UnmountRecursive(target string, flags int) error {
```

Comment:
Should we use https://github.com/moby/sys/blob/c161267cd2212985f81a50a029001f82c91bca7f/mount/mount_unix.go#L41-L44 for this? (and/or see if things are missing in that implementation)

Reply:
Looks like we'll need to refactor `UnmountAll` to `moby/sys/mount` if we want to use that. It doesn't do the repeat unmount till EINVAL.

Reply:
Do we know the reason for the repeat unmount? (I've seen some code where I was wondering: "was this just a big hammer to fix a problem that wasn't fully understood")?

Reply:
Looks like it came from this discussion from PR:
https://github.com/containerd/containerd/pull/1141#discussion_r127015970

Reply:
Thanks for digging that up; yes, that _"feels"_ like a "let's just use a bigger hammer" discussion, and could be worth looking into.

Main motivation for my comment was that I've seen various "subtle" issues with mount/unmount (and more notably; parsing mountinfo). Code like this can easily start spreading around and diverge, and I _think_ it's a worthy investment to have a solid, well-tested, canonical implementation for such things (if possible).

Let me also ping @kolyshkin (who may have some thoughts as well, as he did a lot of digging into subtleties).

I agree with @dmcgowan that that should not be a blocker for this PR though; I think it's fine if we acknowledge this may need to be verified / looked into, and do that separately.